### PR TITLE
Misuv 3192 pta navbar

### PR DESCRIPTION
**Pull request by chandraAndem**

### CODE REVIEW

File: app/controllers/InYearTaxCalculationController.scala [hunk 1]
```diff
@@ -39,23 +39,23 @@ import javax.inject.Inject
...
-                                               auditingService: AuditingService,
-                                               itvcErrorHandler: ItvcErrorHandler,
-                                               implicit val itvcErrorHandlerAgent: AgentItvcErrorHandler,
-                                               val authorisedFunctions: FrontendAuthorisedFunctions,
-                                               val languageUtils: LanguageUtils,
-                                               val retrieveBtaNavBar: BtaNavBarPredicate,
-                                               implicit val appConfig: FrontendAppConfig,
-                                               implicit override val mcc: MessagesControllerComponents,
-                                               implicit val ec: ExecutionContext
+                                                val executionContext: ExecutionContext,
```
Comment (joesheridan) [inline app/controllers/InYearTaxCalculationController.scala:42]: is this new formatting correct?

Comment (chandraAndem) [reply app/controllers/InYearTaxCalculationController.scala:42]: No will reformat it

File: app/controllers/predicates/NavBar.scala [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+import config.featureswitch.FeatureSwitching
+import forms.utils.SessionKeys
+import models.NavBarEnum
+import play.api.i18n.I18nSupport
+import play.api.mvc.Result
+import play.api.mvc.Results.Redirect
+
+import scala.concurrent.Future
+
+trait NavBar extends I18nSupport with FeatureSwitching {
```
Comment (joesheridan) [inline app/controllers/predicates/NavBar.scala:29]: I'd recommend renaming to SaveOrigin or similar

File: app/controllers/predicates/NavBar.scala [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  def saveOriginAndReturnToHomeWithoutQueryParams[A](request: MtdItUserBase[A], navBarFsDisabled: Boolean = true): Future[Result] = {
+    val originStringOpt: Option[String] = request.getQueryString(origin)
+    val redirectToOriginalCall: Future[Result] = Future.successful(Redirect(request.path))
+
+    if (navBarFsDisabled) {
+      redirectToOriginalCall
+    } else {
+      originStringOpt.fold[Future[Result]](ifEmpty = redirectToOriginalCall)(originString =>
+        (NavBarEnum(originString), request.session.get(SessionKeys.origin)) match {
+          case (Some(navBar), Some(sessionOrigin)) if navBar.toString != sessionOrigin =>
```
Comment (joesheridan) [inline app/controllers/predicates/NavBar.scala:44]: recommend changing: Some(navBar) to Some(originStringEnum) or similar

File: app/controllers/predicates/NavBarFromNinoPredicate.scala [hunk 4]
```diff
@@ -49,37 +49,66 @@
...
-          Right(MtdItUserWithNino[A](mtditid = request.mtditid, nino = request.nino, userName = request.userName, btaNavPartial = Some(partial),
-            saUtr = request.saUtr, credId = request.credId, userType = request.userType, arn = request.arn)(request))
+      request.getQueryString(origin).fold[Future[Either[Result, MtdItUserWithNino[A]]]](
+        ifEmpty = retrieveCacheAndHandleNavBar(request))(_ => {
+        saveOriginAndReturnToHomeWithoutQueryParams(request, isDisabled(NavBarFs)).map(Left(_))
+      })
+    }
+  }
+
+  def retrieveCacheAndHandleNavBar[A](request: MtdItUserWithNino[A])
```
Comment (joesheridan) [inline app/controllers/predicates/NavBarFromNinoPredicate.scala:91]: recommend renaming method to retrieveOriginFromSessionAndHandleNavBar

File: app/controllers/predicates/NavBar.scala [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+import play.api.mvc.Result
+import play.api.mvc.Results.Redirect
+
+import scala.concurrent.Future
+
+trait NavBar extends I18nSupport with FeatureSwitching {
+
+  def redirectToHome: Result = Redirect(controllers.routes.HomeController.show())
+
+  lazy val origin: String = "origin"
```
Comment (joesheridan) [inline app/controllers/predicates/NavBar.scala:33]: recommend changing origin to originKey

File: app/controllers/predicates/NavBarPredicate.scala [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    if (isDisabled(NavBarFs)) {
+      Future.successful(Right(request))
+    } else {
+      request.getQueryString(origin).fold[Future[Either[Result, MtdItUser[A]]]](ifEmpty = retrieveCacheAndHandleNavBar(request))(_ => {
+        saveOriginAndReturnToHomeWithoutQueryParams(request, isDisabled(NavBarFs)).map(Left(_))
+      })
+    }
+  }
+
+  def retrieveCacheAndHandleNavBar[A](request: MtdItUser[A])(implicit hc: HeaderCarrier): Future[Either[Result, MtdItUser[A]]] = {
```
Comment (joesheridan) [inline app/controllers/predicates/NavBarPredicate.scala:60]: perhaps we could move these into the Navbar trait and avoid the duplication?

File: app/forms/utils/SessionKeys.scala [hunk 7]
```diff
@@ -23,4 +23,6 @@ object SessionKeys {
   val chargeSummaryBackPage: String = "chargeSummaryBackPage"
 
   val summaryData: String = "summaryData"
+
+  val origin: String = "origin"
```
Comment (joesheridan) [inline app/forms/utils/SessionKeys.scala:27]: if we want this here, we should remove from the trait.

File: app/models/NavBarEnum.scala [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package models
+
+object NavBarEnum extends Enumeration {
```
Comment (joesheridan) [inline app/models/NavBarEnum.scala:19]: recommend changing to OriginEnum

File: app/controllers/predicates/NavBar.scala (hunk 2 above)
Comment (chandraAndem) [reply app/controllers/predicates/NavBar.scala:29]: Changed

File: app/controllers/predicates/NavBar.scala (hunk 3 above)
Comment (chandraAndem) [reply app/controllers/predicates/NavBar.scala:44]: Changes name to originStringEnum.
Also, used enum here to above that conversion to uppercase and compare.

File: app/controllers/predicates/NavBarFromNinoPredicate.scala (hunk 4 above)
Comment (chandraAndem) [reply app/controllers/predicates/NavBarFromNinoPredicate.scala:91]: This was original name before cache, reverted it back.

File: app/controllers/predicates/NavBarPredicate.scala (hunk 6 above)
Comment (chandraAndem) [reply app/controllers/predicates/NavBarPredicate.scala:60]: I tried with return type MtdItUserBase that but both being different return types, I have to match on instance types and return appropriate model back(MdtItUser/MdtItUserWithNino).

File: app/forms/utils/SessionKeys.scala (hunk 7 above)
Comment (chandraAndem) [reply app/forms/utils/SessionKeys.scala:27]: Deleted the trait one.

File: app/models/NavBarEnum.scala (hunk 8 above)
Comment (chandraAndem) [reply app/models/NavBarEnum.scala:19]: Changed

**Review by joesheridan — APPROVED** [review]

**Pull request merged**